### ISSUE TITLE
Plugin is spawning multiple flow processes in different projects

### ISSUE THREAD

**Issue by michalkvasnicak** [amilajack/eslint-plugin-flowtype-errors#27]
When I run eslint configuration with this plugin in different projects there are multiple flow processes running as a result.

Only way to get rid of these processes is to stop them manually.

![image](https://cloud.githubusercontent.com/assets/174716/19044034/e600aa24-8992-11e6-814b-fafb2136dfa1.png)

This is screenshot when I run linting in 2 projects.

Is there a mechanism to stop them automatically? I know that it is hard because you start server on first run and then use `flow check-contents` to keep this plugin performant.

**Comment by michalkvasnicak** [comment]
I've tried to add following line to `collect.js#executeFlow`

```js
process.on('exit', () => childProcess.spawnSync(getFlowBin(), ['stop']));
```

This works nice with webpack in watching mode but is not good if you are using eslint in your editor because it kills flow server everytime lint process is done.

**Comment by amilajack** [comment]
What version are you using? Try both `1.2.0` and `1.3.0` separately. Do you get the same issues with both?

**Comment by michalkvasnicak** [comment]
I am using `1.3.0`. It works good as expected. But problem with multiple servers starts if I open another project where I am using this plugin too. Then it will start another flow server for given project.

I don't see any nice solution to this because we can't detect if you've stopped your work on project. So flow server will stay running.

So how many separate projects you start, that many flow servers you will have running.

For example:

I am working on project A (`/path/a`), start `npm run lint`. This plugin starts flow server and validates source files.
Then I decide to not work on this project and move on separate project B (`/path/b`), start `npm run lint`. This plugin starts flow server for this directory too. There are 2 flow servers running as a result of this operations (each spawning multiple child processes as on screenshot from first post).

Only way to stop server from A is by running `flow stop` in this directory.

**Comment by amilajack** [comment]
What if the servers could have some kind of timeout option where they will automatically be killed after a certain time, say 1 hour? Also how impactful is the instantiation of new flow servers on your machine?

**Comment by michalkvasnicak** [comment]
I haven't noticed any impact on my machine. I came across it when I was debugging my app. It is just eating memory even if it is not used. Timeouts would be good but how should they work? 

For example: I am using webpack in watch mode with `eslint-loader` which runs your plugin. So I can easily detect termination of process and kill flow server too.

Perhaps mentioning it in readme is enough so user will know that they need to stop flow server manually.

**Comment by amilajack** [comment]
I think that this module should be responsible for handling servers that it instantiates. I think the timeout method is a good one.

**Comment by amilajack** [comment]
@jdmota Would be great if you could look into this. I think it can be solved by small json object of when flow was last run on the project and killing the processes if flow hasn't been run on the project in an hour or so?

```js
{
  '/Users/foo/project-1': '1490553978', // <- unix timestamp?
  '/Users/foo/project-2': '1490554013'
}
```

**Comment by jdmota** [comment]
I will take a look. I just need to understand how the flow server works. Maybe trying to start a flow server ahead of time can help... [eslint-plugin-flow-check does that actually.](https://github.com/marionebl/eslint-plugin-flow-check/commit/afc6d059b8c96113feb7460e8b6c442e626220ac)

**Comment by jdmota** [comment]
> I am working on project A (`/path/a`), start `npm run lint`. This plugin starts flow server and validates source files.
Then I decide to not work on this project and move on separate project B (`/path/b`), start `npm run lint`. This plugin starts flow server for this directory too. There are 2 flow servers running as a result of this operations (each spawning multiple child processes as on screenshot from first post).
Only way to stop server from A is by running `flow stop` in this directory.

Probably I'm not understanding the issue correctly... because when I run flow, there are processes that stay on the background, but they close immediately when I close the terminal. So I don't know if there is really a problem that needs to be fixed here...

**Comment by amilajack** [comment]
> but they close immediately when I close the terminal

They actually stay open. This is good because flow can cache. The downside is that we have to kill them after a certain time period

**Comment by jdmota** [comment]
> They actually stay open

In my case, they seem to close... Maybe because I'm on Windows?...

Anyway, how do we kill the processes after a certain time period? Can you explain exactly your ideia? I'm a little confused 😄

**Comment by amilajack** [comment]
You would probably do something like
```js
execSync('flow stop path/to/server')
```

[Here's the documentation](https://flow.org/en/docs/cli/)

**Comment by jdmota** [comment]
Yes. But, that json file you mentioned before; it will only help close processes from previous projects, and it would require someone to run eslint again so that we can check if flow hasn't been run in an hour, am I correct?

(Was I clear? My english is not perfect 😜 )

**Comment by amilajack** [comment]
> so that we can check if flow hasn't been run in an hour, am I correct?

Yes. We may be able to use some kind of process manager like [pm2](https://github.com/Unitech/pm2). I think that will be a much more elegant solution

**Comment by jdmota** [comment]
On the flow documentation I saw that there is a `--from` argument that you can give.

Do you think it can be useful to us? Maybe flow is able to close itself if it knows what started it???

**Comment by amilajack** [comment]
I haven't used it before so I'm not sure. I would recommend asking on the gitter chat or whatever chat the flow team uses now

**Comment by sarkistlt** [comment]
still have this issue, any update or fix?

**Comment by sarkistlt** [comment]
I have only one work around. In my case I'm using webstorm, so I need to turn off Flow type checking from webstorm, remove `flow-bin` from project and install it as global, then it works, but the are a couple problems.

**Comment by amilajack** [comment]
I believe i know the fix for this. But its will take some time, which i dont have a lot of right now. Hope someone can PR this.

**Comment by jdmota** [comment]
Maybe we could add a setting like `stopOnExit`, and if it is true, we do

```js
process.on('exit', () => childProcess.spawnSync(getFlowBin(), ['stop']));
```

I think it would be good, at least as a work around.

Besides, although keeping the flow server alive is good for caching, I think in most cases, when people run `eslint`, they run it once, and expect it to stop. When using this with editors, I suspect the editor actually keeps the process alive (but I could be wrong).

Thoughts?

**Comment by amilajack** [comment]
@jdmota I think we should test this out and see how it impacts performance. How about creating a new branch. I can publish a new prerelease in `v4.0.0-0` so we can test this out

**Comment by sobolevn** [comment]
This is still an issue for me. Maybe there is some workaround?

**Comment by skyrpex** [comment]
In m case, since this plugin uses flow server, other flow plugins like `ide-flowtype` (for Atom) or `Flow Language Support` (for VSCode) won't work.

Do you have similar experiences?

**Comment by christophediprima** [comment]
I have the same issue? Any fix for this?

**Comment by jeffal** [comment]
I've had this issue occur when running multiple versions of `flow-bin` at the same time.

I recommend checking if you have a global version installed or if your repo has multiple `flow-bin` dependencies (monorepo), that they are all in sync across the board.

**Comment by Etheryte** [comment]
For anyone else running into this problem, in our case it was a problem with Flow [running out of memory with its dependency hash table](https://www.karltarvas.com/2021/03/10/eslint-plugin-flowtype-errors-keeps-spawning-and-killing-flow-processes.html). Naturally, no error is logged about it, the server just crashes and is restarted over and over again which results in the above.  

Increasing the hash table size in `.flowconfig` resolved the issue for us:  

```yaml
[options]
server.max_workers=1
sharedmemory.hash_table_pow=22
```
